In [1]:
# Data manipulation imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# SciKit Learn Processing/ Modeling Imports
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn import metrics

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor
from sklearn.svm import SVR
from sklearn.decomposition import PCA

## Import Data
Import data, train-test split, and begin modeling

In [2]:
gov = pd.read_csv('gov.csv')
gov.head()

country   population  area_square_km  density_square_km  \
0          China  1444216.107         9706961           148.7815   
1          India  1393409.038         3287590           423.8391   
2  United States   332915.073         9372610            35.5200   
3      Indonesia   276361.783         1904569           145.1046   
4       Pakistan   225199.937          881912           255.3542   

   average_years_of_school  auto_demo_polity  gdp_per_capita  happiness_index  \
0                     2.77              -7.0      12009.4430            5.140   
1                     1.24               9.0       5722.5186            4.565   
2                    10.61              10.0      53533.2810            7.119   
3                     2.26               9.0       9781.2070            5.399   
4                     1.31               7.0       4815.9243            5.194   

   freedom_index  less_corruption_index  life_expectancy_years  \
0        0.51697                0.02781                 75.928   
1        0.39786                0.08492                 68.607   
2        0.54604                0.15890                 78.910   
3        0.46611                0.00000                 70.768   
4        0.12102                0.10464                 66.577   

   inequality_gini  median_household_income_dollars  \
0             38.5                           6180.0   
1             35.7                           3168.0   
2             41.1                          43585.0   
3             37.8                           2199.0   
4             33.5                           4060.0   

   median_per_capita_income_dollars  median_individual_income_dollars  
0                            1786.0                           10410.0  
1                             616.0                            2130.0  
2                           15480.0                           65760.0  
3                             541.0                            4050.0  
4                             480.0                            1530.0

In [3]:
# Set X and y
X = gov.drop(columns = ['country', 'less_corruption_index'], axis = 1) #drop identifier and target
y = gov['less_corruption_index']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 1)

In [4]:
# Scale data for modeling
ss = StandardScaler()

Z_train = ss.fit_transform(X_train)
Z_test = ss.transform(X_test)

## Modeling
Fit 8 different models with default parameters and then tune the models with the best performing testing scores

In [5]:
np.random.seed(1) #set random seed

# Linear regression model
lr = LinearRegression().fit(Z_train, y_train)
lr_preds_train, lr_preds_test = lr.predict(Z_train), lr.predict(Z_test)

# Ridge model
ridge = Ridge().fit(Z_train, y_train)
ridge_preds_train, ridge_preds_test = ridge.predict(Z_train), ridge.predict(Z_test)

# Lasso model
lasso = Lasso().fit(Z_train, y_train)
lasso_preds_train, lasso_preds_test = lasso.predict(Z_train), lasso.predict(Z_test)

# Random Forest model
rf = RandomForestRegressor(random_state = 1).fit(Z_train, y_train) #set random_state for reprodcucability
rf_preds_train, rf_preds_test = rf.predict(Z_train), rf.predict(X_test)

# KNN model
knn = KNeighborsRegressor().fit(Z_train, y_train)
knn_preds_train, knn_preds_test = knn.predict(Z_train), knn.predict(Z_test)

# Bagging model
bagging = BaggingRegressor().fit(Z_train, y_train)
bagging_preds_train, bagging_preds_test = bagging.predict(Z_train), bagging.predict(Z_test)

# SVR model
svr = SVR().fit(Z_train, y_train)
svr_preds_train, svr_preds_test = svr.predict(Z_train), svr.predict(Z_test)

# PCA model
pca = PCA(random_state = 1).fit(Z_train, y_train)
Z_pca_train, Z_pca_test = pca.transform(Z_train), pca.transform(Z_test)

# Fit on Z_train - credit below to PCA lecture
lr_pca = LinearRegression().fit(Z_pca_train, y_train)
lr_pca_preds_train, lr_pca_Preds_test = lr_pca.predict(Z_train), lr_pca.predict(Z_test)

## Score Models
Check r squared scores for each of the base models - cross validated scores were very low so not used as I think they are an unfair representation of what is happening

In [6]:
# Linear regression scoring
lr_training_score = metrics.r2_score(y_train, lr_preds_train)
lr_testing_score = metrics.r2_score(y_test, lr_preds_test)
print(f'Lr training score: {lr_training_score}')
print(f'Lr testing score: {lr_testing_score} \n')

# Lasso scoring
lasso_training_score = metrics.r2_score(y_train, lasso_preds_train)
lasso_testing_score = metrics.r2_score(y_test, lasso_preds_test)
print(f'Lasso training score: {lasso_training_score}')
print(f'Lasso testing score: {lasso_testing_score} \n')

# Ridge scoring
ridge_training_score = metrics.r2_score(y_train, ridge_preds_train)
ridge_testing_score = metrics.r2_score(y_test, ridge_preds_test)
print(f'Ridge training score: {ridge_training_score}')
print(f'Ridge testing score: {ridge_testing_score} \n')

# Random forest scoring
rf_training_score = metrics.r2_score(y_train, rf_preds_train)
rf_testing_score = metrics.r2_score(y_test, rf_preds_test)
print(f'Rf training score: {rf_training_score}')
print(f'Rf testing score: {rf_testing_score} \n')

# Knn scoring
knn_training_score = metrics.r2_score(y_train, knn_preds_train)
knn_testing_score = metrics.r2_score(y_test, knn_preds_test)
print(f'Knn training score: {knn_training_score}')
print(f'Knn testing score: {knn_testing_score} \n')

# Bagging scoring
bagging_training_score = metrics.r2_score(y_train, bagging_preds_train)
bagging_testing_score = metrics.r2_score(y_test, bagging_preds_test)
print(f'Bagging training score: {bagging_training_score}')
print(f'Bagging testing score: {bagging_testing_score} \n')

# Svr scoring
svr_training_score = metrics.r2_score(y_train, svr_preds_train)
svr_testing_score = metrics.r2_score(y_test, svr_preds_test)
print(f'Svr training score: {svr_training_score}')
print(f'Svr testing score: {svr_testing_score} \n')

# Pca scoring
print(f'Pca training score (lr): {round(lr_pca.score(Z_pca_train, y_train), 4)}')
print(f'Pca testing score (lr): {round(lr_pca.score(Z_pca_test, y_test), 4)}')

Lr training score: 0.6263095320540817
Lr testing score: 0.42903311503631614 

Lasso training score: 0.0
Lasso testing score: -0.022450152580491656 

Ridge training score: 0.6262304029964222
Ridge testing score: 0.42694326454701614 

Rf training score: 0.9262014692963894
Rf testing score: -2.374368779174898 

Knn training score: 0.6328655590543724
Knn testing score: 0.14426616982842 

Bagging training score: 0.9006657005878178
Bagging testing score: 0.4058359959782937 

Svr training score: 0.6728948602561878
Svr testing score: 0.2994719767663966 

Pca training score (lr): 0.6263
Pca testing score (lr): 0.429


## Preliminary Model Perfromance
The three best testing data models were linear regression (.42903), PCA with linear regression (.429), and Ridge (.426). Below parameters are tuned for optimal performance

In [7]:
# Tuned linear regression model gridsearch
lr_params = {
    'fit_intercept' : [True, False],
    'normalize' : [True, False]
}

lr_tuned = GridSearchCV(lr,
                        lr_params,
                        verbose = 1)

lr_tuned = lr_tuned.fit(Z_train, y_train)

Fitting 5 folds for each of 4 candidates, totalling 20 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished


In [8]:
# Score lr_tuned
lr_tuned_preds_train, lr_tuned_preds_test = lr_tuned.predict(Z_train), lr_tuned.predict(Z_test)

lr_tuned_training_score = metrics.r2_score(y_train, lr_tuned_preds_train)
lr_tuned_testing_score = metrics.r2_score(y_test, lr_tuned_preds_test)
print(f'Lr training score: {lr_training_score}')
print(f'Lr testing score: {lr_testing_score} \n')

Lr training score: 0.6263095320540817
Lr testing score: 0.42903311503631614 



No improvement over default parameters

In [9]:
# Tuned PCA model gridsearch
pca_params = {
    'n_components' : [2, 5, None],
    'copy' : [True, False],
    'random_state' : [1],
    'tol' : [0.0, .1, 1]
}

pca_tuned = GridSearchCV(pca,
                         pca_params,
                         cv = 5,
                         verbose = 1)

pca_tuned = pca_tuned.fit(X_train, y_train)
Z_pca_tuned_train = pca_tuned.transform(Z_train)
Z_pca_tuned_test = pca_tuned.transform(Z_test)

# Fit on Z_train - credit below to PCA lecture
lr_pca_tuned = LinearRegression().fit(Z_pca_tuned_train, y_train)
lasso_pca_tuned = Lasso().fit(Z_pca_tuned_train, y_train)
ridge_pca_tuned = Ridge().fit(Z_pca_tuned_train, y_train)
rf_pca_tuned = RandomForestRegressor().fit(Z_pca_tuned_train, y_train)
knn_pca_tuned = KNeighborsRegressor().fit(Z_pca_tuned_train, y_train)
bagging_pca_tuned = BaggingRegressor().fit(Z_pca_tuned_train, y_train)
svr_pca_tuned = SVR().fit(Z_pca_tuned_train, y_train)

Fitting 5 folds for each of 18 candidates, totalling 90 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:    0.4s finished


In [10]:
# Tuned pca scoring - linear regression
print(f'Lr pca tuned training score: {round(lr_pca_tuned.score(Z_pca_tuned_train, y_train), 4)}')
print(f'Lr pca tuned testing score: {round(lr_pca_tuned.score(Z_pca_tuned_test, y_test), 4)} \n')

# Pca scoring - lasso scoring
print(f'Lasso pca tuned training score: {round(lasso_pca_tuned.score(Z_pca_tuned_train, y_train), 4)}')
print(f'Lasso pca tuned testing score: {round(lasso_pca_tuned.score(Z_pca_tuned_test, y_test), 4)} \n')

# Pca scoring - ridge scoring
print(f'Ridge pca tuned training score: {round(ridge_pca_tuned.score(Z_pca_tuned_train, y_train), 4)}')
print(f'Ridge pca tuned testing score: {round(ridge_pca_tuned.score(Z_pca_tuned_test, y_test), 4)} \n')

# Pca scoring - random forest
print(f'Rf pca tuned training score: {round(rf_pca_tuned.score(Z_pca_tuned_train, y_train), 4)}')
print(f'Rf pca tuned testing score: {round(rf_pca_tuned.score(Z_pca_tuned_test, y_test), 4)} \n')

# Pca scoring - knn
print(f'Knn pca tuned training score: {round(knn_pca_tuned.score(Z_pca_tuned_train, y_train), 4)}')
print(f'Knn pca tuned testing score: {round(knn_pca_tuned.score(Z_pca_tuned_test, y_test), 4)} \n')

# Pca scoring - bagging
print(f'Bagging pca tuned training score: {round(bagging_pca_tuned.score(Z_pca_tuned_train, y_train), 4)}')
print(f'Bagging pca tuned testing score: {round(bagging_pca_tuned.score(Z_pca_tuned_test, y_test), 4)} \n')

# Pca scoring - svr
print(f'Svr pca tuned training score: {round(svr_pca_tuned.score(Z_pca_tuned_train, y_train), 4)}')
print(f'Svr pca tuned testing score: {round(svr_pca_tuned.score(Z_pca_tuned_test, y_test), 4)}')

Lr pca tuned training score: 0.6263
Lr pca tuned testing score: 0.429 

Lasso pca tuned training score: 0.0
Lasso pca tuned testing score: -0.0225 

Ridge pca tuned training score: 0.6262
Ridge pca tuned testing score: 0.4269 

Rf pca tuned training score: 0.9249
Rf pca tuned testing score: 0.463 

Knn pca tuned training score: 0.6329
Knn pca tuned testing score: 0.1443 

Bagging pca tuned training score: 0.8681
Bagging pca tuned testing score: 0.3778 

Svr pca tuned training score: -0.001
Svr pca tuned testing score: -0.0134


Rnadom forest turns out to be the best base model for pca

In [11]:
# Tuned ridge model gridsearch
ridge_params = {
    'alpha' : [.01, .1, 1, 10, 100],
    'fit_intercept' : [True, False],
    'normalize' : [True, False],
    'solver' : ['auto', 'svd', 'cholesky', 'lsqr', 'sparse_cg', 'sag', 'saga']
}

ridge_tuned = GridSearchCV(ridge,
                           ridge_params,
                           verbose = 1)

ridge_tuned = ridge_tuned.fit(Z_train, y_train)

# Score ridge_tuned
ridge_tuned_preds_train, ridge_tuned_preds_test = ridge_tuned.predict(Z_train), ridge_tuned.predict(Z_test)

ridge_tuned_training_score = metrics.r2_score(y_train, ridge_tuned_preds_train)
ridge_tuned_testing_score = metrics.r2_score(y_test, ridge_tuned_preds_test)
print(f'Ridge training score: {ridge_training_score}')
print(f'Ridge testing score: {ridge_testing_score} \n')

Fitting 5 folds for each of 140 candidates, totalling 700 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ridge training score: 0.6262304029964222
Ridge testing score: 0.42694326454701614 



[Parallel(n_jobs=1)]: Done 700 out of 700 | elapsed:    1.3s finished


No improvement over default parameters

## Conclusions
There is no improvmenet on the PCA model with linear regression or the gridsearched linear regression and ridge models, but PCA with random forest does perform slightly better with the highest testing score of .463 so this will be the final production model. With more time I would tune a neural networks to try and beat this performance but these results make sense within the context of predicintg for such a dififcult outcome such as corruption. Since principal component analysis does not output meaningful weights for features the best interpretation of results will be correlations from the initial EDA until there is a more involved statistical investigation which is beyond the scope of this study